mlserver/parallel/pool.py

```diff
@@ -0,0 +1,207 @@
+                return
+
+            await self._process_response(response)
+
+    async def _process_response(self, response: InferenceResponseMessage):
+        internal_id = response.id
+
+        async_response = self._async_responses[internal_id]
+
+        if response.inference_response:
+            async_response.set_result(response.inference_response)
+        elif response.exception:
+            async_response.set_exception(response.exception)
+        else:
+            exc = InferenceError("Inference returned no value")
+            async_response.set_exception(exc)
+
+    async def predict(
+        self, model_settings: ModelSettings, inference_request: InferenceRequest
+    ) -> InferenceResponse:
+        internal_id = generate_uuid()
+
+        model_version = None
+        if model_settings.parameters:
+            model_version = model_settings.parameters.version
+
+        request_message = InferenceRequestMessage(
+            id=internal_id,
+            model_name=model_settings.name,
+            model_version=model_version,
+            inference_request=inference_request,
+        )
+        self._requests.put(request_message)
+
+        loop = asyncio.get_running_loop()
+        async_response = loop.create_future()
+        self._async_responses[internal_id] = async_response
+
+        return await self._wait_response(internal_id)
+
+    async def _wait_response(self, internal_id: str) -> InferenceResponse:
+        async_response = self._async_responses[internal_id]
+
+        try:
+            inference_response = await async_response
+            return inference_response
+        finally:
+            del self._async_responses[internal_id]
+
+        return await async_response
+
+    def parallel(self, f: PredictMethod):
+        """
+        Decorator to attach to model's methods so that they run in parallel.
+        By default, this will get attached to every model's "inference" method.
+
+        NOTE: At the moment, this method only works with `predict()`.
+        """
+        # TODO: Extend to multiple methods
+        @wraps(f)
+        async def _inner(payload: InferenceRequest) -> InferenceResponse:
+            wrapped_f = get_wrapped_method(f)
+            if not hasattr(wrapped_f, "__self__"):
+                raise InvalidParallelMethod(
+                    wrapped_f.__name__, reason="method is not bound"
+                )
+
+            model = getattr(wrapped_f, "__self__")
+
+            return await self.predict(model.settings, payload)
+
+        return _inner
+
+    async def load_model(self, model: MLModel):
+        if not self._should_load_model(model):
+            # Skip load if model has disabled parallel workers
+            return
+
+        load_message = ModelUpdateMessage(
+            update_type=ModelUpdateType.Load, model_settings=model.settings
+        )
+        await asyncio.gather(
+            *[worker.send_update(load_message) for worker in self._workers.values()]
```

Comment:
It's just a helper to hide some implementation details away from the pool. I agree it's confusing because it will run on the pool's main process, rather than the worker, but I felt like the pool was already quite overcrowded.